### DISCUSSION TITLE
Indentation of the `then` statement in `case` expressions

### DISCUSSION THREAD

**Discussion by f-tremblay** [tconbeer/sqlfmt#303] [The sqlfmt Style]
**Describe the bug**
In `case` expressions, when there are long `when` statements, sqlfmt put the `then` statements at the same indentation as the `when` statements, which makes them harder to read.
If we look at the dbt style guide (e.g. [here](https://github.com/dbt-labs/corp/blob/main/dbt_style_guide.md#sql-style-guide) and [here](https://github.com/dbt-labs/corp/blob/main/dbt_style_guide.md#example-sql)), the `then` is more indented.

**To Reproduce**
  ```sql
with
    source as (select * from {{ source("my_application", "users") }}),
    renamed as (
        select
            sum(
                case when order_status = 'complete' then order_total end
            ) as monthly_total,
            nullif(
                trim(
                    case
                        when regexp_count(nullif(full_name, ''), ' ') = 0
                        then nullif(full_name, '')
                        when regexp_count(nullif(full_name, ''), ' ') = 1
                        then split_part(nullif(full_name, ''), ' ', 1)
                        -- this user has more than two names!
                        else regexp_substr(nullif(full_name, ''), '.* .* ')
                    end
                )

                from source
            )
        select *
        from renamed
  ```

**Expected behavior**
  ```sql
with
    source as (select * from {{ source("my_application", "users") }}),
    renamed as (
        select
            sum(
                case when order_status = 'complete' then order_total end
            ) as monthly_total,
            nullif(
                trim(
                    case
                        when regexp_count(nullif(full_name, ''), ' ') = 0
                            then nullif(full_name, '')
                        when regexp_count(nullif(full_name, ''), ' ') = 1
                            then split_part(nullif(full_name, ''), ' ', 1)
                        -- this user has more than two names!
                        else regexp_substr(nullif(full_name, ''), '.* .* ')
                    end
                )

                from source
            )
        select *
        from renamed
  ```

**Actual behavior**
`then`are not indented properly (see `To Reproduce` )

**Additional context**
What is the output of `sqlfmt --version`?
sqlfmt, version 0.13.0

**Comment by tconbeer** [comment]
@f-tremblay Thanks for the feedback. This is the intended behavior, so I've moved this to a discussion.

I think indented `then` (like `join`) has its merits, but I intentionally diverged from the dbt style guide for the following reasons:

1. Keeping `when`, `then`, and `else` at the same depth makes the code much simpler
2. There will always be collisions with `then`. If you indent `then`, it can blend in with complex conditionals:
    ```sql
    case
        when
            fooooooo.barrrrrrr = another_table.another_column
            and fooooooo.barrrrrrr = another_table.another_column2
            then  fooooooo.barrrrrrr + another_table.another_column
    ...
    ```

3. What about `else`? The code after `else` is analagous to what follows `then`, not `when`. So shouldn't `else` have the same indentation as `then`? That doesn't make sense if you indent `then`, since `else` will be nested under the final `when`. This is especially incongruous when the statements might wrap onto a single line (I wouldn't want `when` and `then` on a single line and `else` on a following line).

**Reply by f-tremblay** [reply]
Thanks @tconbeer for the explanations.

For the first point, were you talking about sqlfmt's code, or the user's SQL code?

**Reply by f-tremblay** [reply]
For the second point, I will use python as an example to share my thoughts. 

```python
if foooooo == 1 and baaaaar == 1:
    out = "foooooobaaaaar"
elif foooooo == 1:
    out = "foooooo"
```
and
```sql
case
    when foooooo = 1 and baaaaar = 1
        then "foooooobaaaaar"
    when foooooo = 1
        then "foooooo"
end as out,
```
are both easy to read.

However, 
```python
if foooooo == 1 and baaaaar == 1:
out = "foooooobaaaaar"
elif foooooo == 1:
out = "foooooo"
```
is incorrect, and 
```sql
case
    when foooooo = 1 and baaaaar = 1
    then "foooooobaaaaar"
    when foooooo = 1
    then "foooooo"
end as out,
```
is painful to read. 
(Since the words `when` and `then` are very similar, without proper indentation, it requires extra attention not to confuse them)

These principles still apply for more complex (i.e. longer) conditions:
```python
if (
    foooooo == 1
    and baaaaar == 1
    and another_longer_condition == "a very very long string"
):
    out = "foooooobaaaaar"
elif foooooo == 1:
    out = "foooooo"
```
and
```sql
case
    when
        foooooo = 1
        and baaaaar = 1
        and another_longer_condition == "a very very long string"
        then "foooooobaaaaar"
    when foooooo = 1
        then "foooooo"
end as out,
```
are both easy to read.

However,
```python
if (
    foooooo == 1
    and baaaaar == 1
    and another_longer_condition == "a very very long string"
):
out = "foooooobaaaaar"
elif foooooo == 1:
out = "foooooo"
```
is incorrect, and
```sql
case
    when
        foooooo = 1
        and baaaaar = 1
        and another_longer_condition == "a very very long string"
    then "foooooobaaaaar"
    when foooooo = 1
    then "foooooo"
end as out,
```
is very confusing

**Reply by f-tremblay** [reply]
For the last point, I will also use python:

```python
if foooooo == 1 and baaaaar == 1:
    out = "foooooobaaaaar"
elif foooooo == 1:
    out = "foooooo"
else: out = ""
```
is easy to read, although 
```python
if foooooo == 1 and baaaaar == 1:
    out = "foooooobaaaaar"
elif foooooo == 1:
    out = "foooooo"
else: 
    out = ""
```
is preferable according to many (including black).
Similarly,
```sql
case
    when foooooo = 1 and baaaaar = 1
        then "foooooobaaaaar"
    when foooooo = 1
        then "foooooo"
    else ""
end as out,
```
is easy to read, although 
```sql
case
    when foooooo = 1 and baaaaar = 1
        then "foooooobaaaaar"
    when foooooo = 1
        then "foooooo"
    else 
        ""
end as out,
```
would be equally readable.


However, 
```python
if foooooo == 1 and baaaaar == 1:
out = "foooooobaaaaar"
elif foooooo == 1:
out = "foooooo"
else:
out = ""
```
is incorrect, and 
```sql
case
    when foooooo = 1 and baaaaar = 1
    then "foooooobaaaaar"
    when foooooo = 1
    then "foooooo"
    else ""
end as out,
```
is confusing. 

It simply does not make sense to have `then` at the same indentation as `when` and `else`.  `when` and `else` both indicate that there's a new "case" (the number of "cases" is the number of  `when` +`else`  keyword(s) in a case expression), while the `then` keyword is to separate the condition(s) from the result expression.
